From 41924113f31403d7bc67df9df2d213ffd742ea21 Mon Sep 17 00:00:00 2001
From: Koichi Murase <myoga.murase@gmail.com>
Date: Thu, 3 Nov 2022 14:01:41 +0900
Subject: [PATCH 2/3] fix memleak in parameter_brace_expand_word

---
 subst.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/subst.c b/subst.c
index 7ec8f8e9..b62ebc6c 100644
--- a/subst.c
+++ b/subst.c
@@ -7517,7 +7517,7 @@ expand_arrayref:
       else if (es.subtype == 2 && temp && QUOTED_NULL (temp) && (quoted & (Q_DOUBLE_QUOTES|Q_HERE_DOCUMENT)))
 	rflags |= W_HASQUOTEDNULL;
 
-      if (estatep == 0)
+      if (!(es.subtype == 0 && temp && estatep))
 	flush_eltstate (&es);
     }
 #endif
-- 
2.37.2

