From e427003f81c0283ce77a18fa565191aed1845a72 Mon Sep 17 00:00:00 2001
From: Koichi Murase <myoga.murase@gmail.com>
Date: Fri, 24 Jan 2020 07:19:49 +0800
Subject: [PATCH 8/8] remove stdout.on/off

---
 ble.sh | 92 ----------------------------------------------------------
 1 file changed, 92 deletions(-)

diff --git a/ble.sh b/ble.sh
index 0c40c0f..0a42f4d 100644
--- a/ble.sh
+++ b/ble.sh
@@ -13534,98 +13534,6 @@ function ble/widget/command-help {
 function ble-edit/bind/stdout.on { :;}
 function ble-edit/bind/stdout.off { ble/util/buffer.flush >&2;}
 function ble-edit/bind/stdout.finalize { :;}
-if [[ $bleopt_internal_suppress_bash_output ]]; then
-  _ble_edit_io_stdout=
-  _ble_edit_io_stderr=
-  ble/util/openat _ble_edit_io_stdout '>&1'
-  ble/util/openat _ble_edit_io_stderr '>&2'
-  _ble_edit_io_fname1=$_ble_base_run/$$.stdout
-  _ble_edit_io_fname2=$_ble_base_run/$$.stderr
-  function ble-edit/bind/stdout.on {
-    exec 1>&$_ble_edit_io_stdout 2>&$_ble_edit_io_stderr
-  }
-  function ble-edit/bind/stdout.off {
-    ble/util/buffer.flush >&2
-    ble-edit/bind/stdout/check-stderr
-    exec 1>>$_ble_edit_io_fname1 2>>$_ble_edit_io_fname2
-  }
-  function ble-edit/bind/stdout.finalize {
-    ble-edit/bind/stdout.on
-    [[ -f $_ble_edit_io_fname1 ]] && ble/bin/rm -f "$_ble_edit_io_fname1"
-    [[ -f $_ble_edit_io_fname2 ]] && ble/bin/rm -f "$_ble_edit_io_fname2"
-  }
-  function ble-edit/bind/stdout/check-stderr {
-    local file=${1:-$_ble_edit_io_fname2}
-    if ble/is-function ble/term/visible-bell; then
-      if [[ -f $file && -s $file ]]; then
-        local message= line
-        while IFS= builtin read -r line || [[ $line ]]; do
-          if [[ $line == 'bash: '* || $line == "${BASH##*/}: "* ]]; then
-            message="$message${message:+; }$line"
-          fi
-        done < "$file"
-        [[ $message ]] && ble/term/visible-bell "$message"
-        : >| "$file"
-      fi
-    fi
-  }
-  if ((_ble_bash<40000)); then
-    function ble-edit/bind/stdout/TRAPUSR1 {
-      [[ $_ble_term_state == internal ]] || return
-      local FUNCNEST=
-      local IFS=$' \t\n'
-      local file=$_ble_edit_io_fname2.proc
-      if [[ -s $file ]]; then
-        local content cmd
-        ble/util/readfile content "$file"
-        : >| "$file"
-        for cmd in $content; do
-          case "$cmd" in
-          (eof)
-            ble-decode/.hook 4
-            builtin eval "$_ble_decode_bind_hook" ;;
-          esac
-        done
-      fi
-      ble/builtin/trap/invoke USR1
-    }
-    ble/builtin/trap/reserve USR1
-    builtin trap -- 'ble-edit/bind/stdout/TRAPUSR1' USR1
-    ble/bin/rm -f "$_ble_edit_io_fname2.pipe"
-    ble/bin/mkfifo "$_ble_edit_io_fname2.pipe"
-    {
-      {
-        function ble-edit/stdout/check-ignoreeof-message {
-          local line=$1
-          [[ $line == *$bleopt_internal_ignoreeof_trap* ||
-               $line == *'Use "exit" to leave the shell.'* ||
-               $line == *'ログアウトする為には exit を入力して下さい'* ||
-               $line == *'シェルから脱出するには "exit" を使用してください。'* ||
-               $line == *'シェルから脱出するのに "exit" を使いなさい.'* ||
-               $line == *'Gebruik Kaart na Los Tronk'* ]] && return 0
-          [[ $line == *exit* ]] && ble/bin/grep -q -F "$line" "$_ble_base"/lib/core-edit.ignoreeof-messages.txt
-        }
-        while IFS= builtin read -r line; do
-          SPACE=$' \n\t'
-          if [[ $line == *[^$SPACE]* ]]; then
-            builtin printf '%s\n' "$line" >> "$_ble_edit_io_fname2"
-          fi
-          if [[ $bleopt_internal_ignoreeof_trap ]] && ble-edit/stdout/check-ignoreeof-message "$line"; then
-            ble/util/print eof >> "$_ble_edit_io_fname2.proc"
-            kill -USR1 $$
-            ble/util/msleep 100 # 連続で送ると bash が落ちるかも (落ちた事はないが念の為)
-          fi
-        done < "$_ble_edit_io_fname2.pipe"
-      } &>/dev/null & disown
-    } &>/dev/null
-    ble/util/openat _ble_edit_fd_stderr_pipe '> "$_ble_edit_io_fname2.pipe"'
-    function ble-edit/bind/stdout.off {
-      ble/util/buffer.flush >&2
-      ble-edit/bind/stdout/check-stderr
-      exec 1>>$_ble_edit_io_fname1 2>&$_ble_edit_fd_stderr_pipe
-    }
-  fi
-fi
 [[ $_ble_edit_detach_flag != reload ]] &&
   _ble_edit_detach_flag=
 function ble-edit/bind/.exit-TRAPRTMAX {
-- 
2.21.0

