From 0bb8d448aa628e240c78028ea362203854e30409 Mon Sep 17 00:00:00 2001
From: Koichi Murase <myoga.murase@gmail.com>
Date: Mon, 11 Feb 2019 15:27:58 +0900
Subject: [PATCH 2/3] add a range condition to "rl_bind_key"

---
 lib/readline/bind.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/readline/bind.c b/lib/readline/bind.c
index 042c345..50f3109 100644
--- a/lib/readline/bind.c
+++ b/lib/readline/bind.c
@@ -138,7 +138,7 @@ rl_bind_key (int key, rl_command_func_t *function)
   char keyseq[3];
   int l;
 
-  if (key < 0)
+  if (key < 0 || key > largest_char)
     return (key);
 
   if (META_CHAR (key) && _rl_convert_meta_chars_to_ascii)
-- 
2.9.5

