From 475e292ae83bb516cc8d33f0f3ccc1a9b4b4f55b Mon Sep 17 00:00:00 2001
From: Koichi Murase <myoga.murase@gmail.com>
Date: Tue, 23 Jul 2019 06:44:38 +0900
Subject: [PATCH 3/3] Improve the initial value for "do_and"

---
 builtin.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/builtin.c b/builtin.c
index 5f62426d..1c205aa4 100644
--- a/builtin.c
+++ b/builtin.c
@@ -3468,7 +3468,7 @@ do_and(int nargs)
 	uintmax_t res, uval;
 	AWKNUM val;
 
-	res = ~0;	/* start off with all ones */
+	res = ~(uintmax_t) 0;	/* start off with all ones */
 	if (nargs < 2)
 		fatal(_("and: called with less than two arguments"));
 
-- 
2.21.0

